File: src/exit_manager.py
"""Exit manager - monitor positions and generate exit signals.

Responsibilities:
- Monitor open positions periodically (every check_interval_seconds)
- Calculate P&L for each position against entry price
- Evaluate exit rules: stop loss (-1%), profit target (+2%), trailing stop (optional)
- Generate ExitSignalEvent when thresholds breached
- Track trailing stop levels per position
- Provide close_all_positions() for emergency exits

Exit Rules Priority:
1. Stop loss (highest priority) - IF unrealised_pnl_pct <= -stop_loss_pct
2. Trailing stop (if activated) - IF current_price <= trailing_stop_price
3. Profit target - IF unrealised_pnl_pct >= profit_target_pct
"""

import asyncio
import logging
import math
from datetime import datetime, timezone
from typing import Any, Optional

from src.broker import Broker
from src.data_handler import DataHandler
from src.event_bus import EventBus, ExitSignalEvent
from src.position_tracker import PositionData, PositionTracker
from src.state_store import StateStore

logger = logging.getLogger(__name__)


def calculate_dynamic_stops(
    entry_price: float,
    atr: float,
    atr_multiplier_stop: float = 1.5,
    atr_multiplier_target: float = 3.0,
    side: str = "long",
) -> tuple[float, float]:
    """Calculate stop loss and profit target based on ATR.

    Returns `(stop_price, target_price)` for the given `side`.

    For `long`:
      - stop_price = entry_price - stop_distance
      - target_price = entry_price + target_distance

    For `short`:
      - stop_price = entry_price + stop_distance
      - target_price = entry_price - target_distance
    """
    stop_distance = atr * atr_multiplier_stop
    target_distance = atr * atr_multiplier_target

    converted_side = (side or "").lower()
    # Validate side explicitly to avoid silently treating unknown values as `long`.
    if converted_side not in ("long", "short"):
        raise ValueError(f"Invalid side for dynamic stop calculation: {side!r}")

    if converted_side == "short":
        stop_price = entry_price + stop_distance
        target_price = entry_price - target_distance
    else:
        stop_price = entry_price - stop_distance
        target_price = entry_price + target_distance

    return stop_price, target_price


class ExitManagerError(Exception):
    """Raised when exit manager operation fails."""

    pass


class ExitManager:
    """Monitor positions and generate exit signals."""

    def __init__(
        self,
        broker: Broker,
        position_tracker: PositionTracker,
        event_bus: EventBus,
        state_store: StateStore,
        data_handler: DataHandler,
        stop_loss_pct: float = 0.01,
        profit_target_pct: float = 0.02,
        trailing_stop_enabled: bool = False,
        trailing_stop_activation_pct: float = 0.01,
        trailing_stop_trail_pct: float = 0.005,
        check_interval_seconds: int = 30,
        exit_on_circuit_breaker: bool = True,
        atr_multiplier_stop: float = 1.5,
        atr_multiplier_target: float = 3.0,
    ) -> None:
        """Initialise exit manager.

        Args:
            broker: Broker client for price fetching
            position_tracker: Position tracker for entry prices
            event_bus: Event bus for publishing exit signals
            state_store: State store for persistence
            data_handler: Data handler for price data
            stop_loss_pct: Stop loss percentage (e.g., 0.01 = -1%)
            profit_target_pct: Profit target percentage (e.g., 0.02 = +2%)
            trailing_stop_enabled: Whether trailing stops are enabled
            trailing_stop_activation_pct: P&L % to activate trailing stop
            trailing_stop_trail_pct: Distance below highest price for trailing stop
            check_interval_seconds: How often to check positions
            exit_on_circuit_breaker: Whether to close all positions on circuit breaker
        """
        self.broker = broker
        self.position_tracker = position_tracker
        self.event_bus = event_bus
        self.state_store = state_store
        self.data_handler = data_handler

        self.stop_loss_pct = stop_loss_pct
        self.profit_target_pct = profit_target_pct
        self.trailing_stop_enabled = trailing_stop_enabled
        self.trailing_stop_activation_pct = trailing_stop_activation_pct
        self.trailing_stop_trail_pct = trailing_stop_trail_pct
        self.check_interval_seconds = check_interval_seconds
        self.exit_on_circuit_breaker = exit_on_circuit_breaker

        # ATR multipliers for dynamic stops/targets
        self.atr_multiplier_stop = atr_multiplier_stop
        self.atr_multiplier_target = atr_multiplier_target

        self._running = False
        self._monitor_task: Optional[asyncio.Task[Any]] = None

    async def start(self) -> None:
        """Start the exit manager monitoring loop."""
        if self._running:
            logger.warning("Exit manager already running")
            return

        logger.info("Starting exit manager...")
        logger.info(
            f"  Stop loss: -{self.stop_loss_pct*100:.1f}%, "
            f"Profit target: +{self.profit_target_pct*100:.1f}%"
        )
        logger.info(
            f"  ATR multipliers: stop={self.atr_multiplier_stop}, target={self.atr_multiplier_target}"
        )
        if self.trailing_stop_enabled:
            logger.info(
                f"  Trailing stop: enabled (activation +{self.trailing_stop_activation_pct*100:.1f}%, "
                f"trail -{self.trailing_stop_trail_pct*100:.1f}%)"
            )
        else:
            logger.info("  Trailing stop: disabled")

        self._running = True
        self._monitor_task = asyncio.create_task(self._monitor_loop())
        logger.info(f"Exit manager started (checking every {self.check_interval_seconds}s)")

    async def stop(self) -> None:
        """Stop the exit manager monitoring loop."""
        if not self._running:
            return

        logger.info("Stopping exit manager...")
        self._running = False

        if self._monitor_task:
            self._monitor_task.cancel()
            try:
                await self._monitor_task
            except asyncio.CancelledError:
                pass

        logger.info("Exit manager stopped")

    async def _monitor_loop(self) -> None:
        """Main monitoring loop."""
        logger.info("Exit monitor loop started")

        while self._running:
            try:
                # Check circuit breaker first
                if self.exit_on_circuit_breaker:
                    cb_state = self.state_store.get_state("circuit_breaker_state")
                    if cb_state == "tripped":
                        logger.warning("Circuit breaker tripped - closing all positions")
                        await self.close_all_positions(reason="circuit_breaker")
                        await asyncio.sleep(self.check_interval_seconds)
                        continue

                # Check positions
                await self.check_positions()

            except Exception as e:
                logger.error(f"Error in exit monitor loop: {e}", exc_info=True)

            try:
                await asyncio.sleep(self.check_interval_seconds)
            except asyncio.CancelledError:
                break

        logger.info("Exit monitor loop ended")

    async def check_positions(self) -> list[ExitSignalEvent]:
        """Check all tracked positions for exit conditions.

        Returns:
            List of ExitSignalEvents generated
        """
        signals: list[ExitSignalEvent] = []
        positions = self.position_tracker.get_all_positions()

        if not positions:
            return signals

        # Check if market is open first (fresh clock call)
        try:
            clock = self.broker.get_clock()
            if not clock["is_open"]:
                logger.debug("Market closed - skipping position checks")
                return signals
        except Exception as e:
            logger.error(f"Failed to check market clock: {e}")
            return signals

        for position in positions:
            try:
                # Skip if exit signal already pending (deduplication)
                if position.pending_exit:
                    logger.debug(f"Skipping {position.symbol} - exit signal pending")
                    continue

                # Get current price
                snapshot = self.data_handler.get_snapshot(position.symbol)
                if not snapshot:
                    logger.warning(f"No snapshot available for {position.symbol}")
                    continue

                current_price = snapshot.get("last_price") or snapshot.get("bid")
                if not current_price:
                    logger.warning(f"No current price available for {position.symbol}")
                    continue

                # Update position with current price (for trailing stop)
                self.position_tracker.update_current_price(position.symbol, current_price)

                # Check for exit signal
                signal = self._evaluate_exit_rules(position, current_price)
                if signal:
                    signals.append(signal)
                    # Publish exit signal BEFORE marking pending_exit to avoid
                    # persisting a stuck pending_exit if publishing fails.
                    try:
                        await self.event_bus.publish(signal)
                    except Exception as e:
                        logger.error(f"Failed to publish exit signal for {position.symbol}: {e}")
                        # Do not mark pending_exit so the position can be retried
                        continue
                    # Mark position as having pending exit only after successful publish
                    position.pending_exit = True
                    self.position_tracker._persist_position(position)
                    logger.info(
                        f"Exit signal: {signal.symbol} {signal.reason} "
                        f"(P&L: {signal.pnl_pct*100:.1f}%)"
                    )

            except Exception as e:
                logger.error(f"Error checking position {position.symbol}: {e}")

        return signals

    def _evaluate_exit_rules(
        self,
        position: PositionData,
        current_price: float,
    ) -> Optional[ExitSignalEvent]:
        """Evaluate exit rules for a position.

        Exit Rules Priority:
        1. Stop loss (highest priority) - IF unrealised_pnl_pct <= -stop_loss_pct
        2. Trailing stop (if activated) - IF current_price <= trailing_stop_price
        3. Profit target - IF unrealised_pnl_pct >= profit_target_pct

        Args:
            position: PositionData
            current_price: Current market price

        Returns:
            ExitSignalEvent if threshold breached, None otherwise
        """
        # Calculate P&L
        pnl_amount, pnl_pct = self.position_tracker.calculate_pnl(position.symbol, current_price)

        # Whether ATR-based thresholds were computed and applied (skip fallbacks)
        atr_computed = False

        # If ATR is available for this position, use ATR-based dynamic stops/targets
        if position.atr is not None and math.isfinite(position.atr) and position.atr > 0.0:
            atr_value = position.atr
            converted_side = (position.side or "").lower()

            # Validate side explicitly to avoid mismatched compute vs checks
            if converted_side not in ("long", "short"):
                logger.warning(
                    "Unsupported position side '%s' for ATR-based exits on %s; skipping ATR logic.",
                    position.side,
                    position.symbol,
                )
            else:
                stop_price, target_price = calculate_dynamic_stops(
                    entry_price=position.entry_price,
                    atr=atr_value,
                    atr_multiplier_stop=self.atr_multiplier_stop,
                    atr_multiplier_target=self.atr_multiplier_target,
                    side=converted_side,
                )

                # Basic validation of computed stop/target (ensure finite numbers)
                if not (
                    isinstance(stop_price, (int, float)) and isinstance(target_price, (int, float))
                ) or not (math.isfinite(stop_price) and math.isfinite(target_price)):
                    logger.debug(
                        "Computed non-finite stop/target for %s (atr=%s) - skipping ATR logic",
                        position.symbol,
                        atr_value,
                    )
                else:
                    # Mark that ATR-based thresholds were computed; if they are
                    # valid we prefer them over fallback fixed-percentage rules.
                    atr_computed = True
                    # Long position ATR-based checks
                    if converted_side == "long":
                        if current_price <= stop_price:
                            return ExitSignalEvent(
                                symbol=position.symbol,
                                side="sell",
                                qty=position.qty,
                                reason="stop_loss",
                                entry_price=position.entry_price,
                                current_price=current_price,
                                pnl_pct=pnl_pct,
                                pnl_amount=pnl_amount,
                                timestamp=datetime.now(timezone.utc),
                            )

                        if current_price >= target_price:
                            return ExitSignalEvent(
                                symbol=position.symbol,
                                side="sell",
                                qty=position.qty,
                                reason="profit_target",
                                entry_price=position.entry_price,
                                current_price=current_price,
                                pnl_pct=pnl_pct,
                                pnl_amount=pnl_amount,
                                timestamp=datetime.now(timezone.utc),
                            )
                    elif converted_side == "short":
                        # Short position ATR-based checks (inverse)
                        if current_price >= stop_price:
                            return ExitSignalEvent(
                                symbol=position.symbol,
                                side="buy",
                                qty=position.qty,
                                reason="stop_loss",
                                entry_price=position.entry_price,
                                current_price=current_price,
                                pnl_pct=pnl_pct,
                                pnl_amount=pnl_amount,
                                timestamp=datetime.now(timezone.utc),
                            )

                        if current_price <= target_price:
                            return ExitSignalEvent(
                                symbol=position.symbol,
                                side="buy",
                                qty=position.qty,
                                reason="profit_target",
                                entry_price=position.entry_price,
                                current_price=current_price,
                                pnl_pct=pnl_pct,
                                pnl_amount=pnl_amount,
                                timestamp=datetime.now(timezone.utc),
                            )

        # Priority 1: Stop loss (highest priority) - fallback to fixed pct
        if not atr_computed and pnl_pct <= -self.stop_loss_pct:
            return ExitSignalEvent(
                symbol=position.symbol,
                side="sell" if position.side == "long" else "buy",
                qty=position.qty,
                reason="stop_loss",
                entry_price=position.entry_price,
                current_price=current_price,
                pnl_pct=pnl_pct,
                pnl_amount=pnl_amount,
                timestamp=datetime.now(timezone.utc),
            )

        # Priority 2: Trailing stop (if activated and enabled)
        if self.trailing_stop_enabled and position.trailing_stop_activated:
            if (
                position.trailing_stop_price is not None
                and isinstance(position.trailing_stop_price, (int, float))
                and math.isfinite(position.trailing_stop_price)
                and current_price <= position.trailing_stop_price
            ):
                return ExitSignalEvent(
                    symbol=position.symbol,
                    side="sell" if position.side == "long" else "buy",
                    qty=position.qty,
                    reason="trailing_stop",
                    entry_price=position.entry_price,
                    current_price=current_price,
                    pnl_pct=pnl_pct,
                    pnl_amount=pnl_amount,
                    timestamp=datetime.now(timezone.utc),
                )

        # Priority 3: Profit target
        if not atr_computed and pnl_pct >= self.profit_target_pct:
            return ExitSignalEvent(
                symbol=position.symbol,
                side="sell" if position.side == "long" else "buy",
                qty=position.qty,
                reason="profit_target",
                entry_price=position.entry_price,
                current_price=current_price,
                pnl_pct=pnl_pct,
                pnl_amount=pnl_amount,
                timestamp=datetime.now(timezone.utc),
            )

        return None

    async def close_all_positions(self, reason: str = "emergency") -> list[ExitSignalEvent]:
        """Close all open positions immediately.

        Generates ExitSignalEvent for each position.

        Args:
            reason: Reason for closing (e.g., "emergency", "circuit_breaker", "shutdown")

        Returns:
            List of ExitSignalEvents generated
        """
        logger.warning(f"Closing all positions (reason: {reason})")

        signals = []
        positions = self.position_tracker.get_all_positions()

        for position in positions:
            try:
                # Get current price
                snapshot = self.data_handler.get_snapshot(position.symbol)
                if snapshot is None:
                    current_price = position.entry_price
                else:
                    current_price = (
                        snapshot.get("last_price") or snapshot.get("bid") or position.entry_price
                    )

                # Calculate P&L
                pnl_amount, pnl_pct = self.position_tracker.calculate_pnl(
                    position.symbol, current_price
                )

                signal = ExitSignalEvent(
                    symbol=position.symbol,
                    side="sell" if position.side == "long" else "buy",
                    qty=position.qty,
                    reason=reason,
                    entry_price=position.entry_price,
                    current_price=current_price,
                    pnl_pct=pnl_pct,
                    pnl_amount=pnl_amount,
                    timestamp=datetime.now(timezone.utc),
                )

                signals.append(signal)
                await self.event_bus.publish(signal)
                logger.info(
                    f"Emergency exit: {signal.symbol} {signal.qty} shares "
                    f"(P&L: {signal.pnl_pct*100:.1f}%)"
                )

            except Exception as e:
                logger.error(f"Error generating exit signal for {position.symbol}: {e}")

        logger.warning(f"Generated {len(signals)} exit signals for {reason}")
        return signals

    async def check_single_position(self, symbol: str) -> Optional[ExitSignalEvent]:
        """Check a single position for exit conditions."""
        """Check a single position for exit conditions.

        Args:
            symbol: Stock symbol to check

        Returns:
            ExitSignalEvent if threshold breached, None otherwise
        """
        position = self.position_tracker.get_position(symbol)
        if not position:
            return None

        # Get current price
        snapshot = self.data_handler.get_snapshot(symbol)
        if not snapshot:
            return None

        current_price: Optional[float] = snapshot.get("last_price") or snapshot.get("bid")
        if not current_price:
            return None

        # Update position with current price
        self.position_tracker.update_current_price(symbol, current_price)

        # Evaluate exit rules
        return self._evaluate_exit_rules(position, current_price)

File: cs/src/AlpacaFleece.Trading/Exits/ExitManager.cs
using AlpacaFleece.Infrastructure.Symbols;

namespace AlpacaFleece.Trading.Exits;

/// <summary>
/// Exit manager: checks positions periodically for stop loss, trailing stop, profit target.
/// Uses a 3-rule priority system with ATR-based dynamic levels (mutual exclusion with fixed-%):
///
///   Rule 1: ATR stop loss     — entry - (ATR x AtrStopLossMultiplier)
///   Rule 2: ATR profit target — entry + (ATR x AtrProfitTargetMultiplier)
///   Rule 3: Trailing stop     — TrailingStopPrice (always active)
///
/// ATR mutual exclusion: when valid ATR levels exist, fixed-percentage rules are skipped.
/// Publishes ExitSignalEvent to unbounded event bus (never dropped).
/// PendingExit flag is set AFTER successful publish to avoid phantom locks on bus failures.
/// </summary>
/// <param name="positionTracker">The position tracker for accessing open positions.</param>
/// <param name="brokerService">The broker service for executing exit trades.</param>
/// <param name="marketDataClient">The market data client for fetching prices.</param>
/// <param name="eventBus">The event bus for publishing exit signals.</param>
/// <param name="stateRepository">The state repository for storing exit attempt records.</param>
/// <param name="logger">The logger instance.</param>
/// <param name="options">The trading options configuration.</param>
/// <param name="symbolClassifier">Optional symbol classifier for crypto/equity detection.</param>
/// <param name="volatilityRegimeDetector">Optional volatility regime detector for ATR distance adaptation.</param>
public class ExitManager(
    IPositionTracker positionTracker,
    IBrokerService brokerService,
    IMarketDataClient marketDataClient,
    IEventBus eventBus,
    IStateRepository stateRepository,
    ILogger<ExitManager> logger,
    IOptions<TradingOptions> options,
    ISymbolClassifier? symbolClassifier = null,
    VolatilityRegimeDetector? volatilityRegimeDetector = null)
{
    /// <summary>
    /// Protected no-arg constructor for NSubstitute proxy creation in testing.
    /// </summary>
    protected ExitManager() : this(null!, null!, null!, null!, null!, null!, Options.Create(new TradingOptions()), null!, null!) { }

    private readonly ExitOptions _options = options.Value.Exit;
    private readonly ISymbolClassifier _symbolClassifier = symbolClassifier ?? new SymbolClassifier(options.Value.Symbols.CryptoSymbols, options.Value.Symbols.EquitySymbols);

    // Per-symbol consecutive price-fetch failure counter.
    // After MaxPriceFailures consecutive failures the market_data_degraded KV is set,
    // which blocks new entries in RiskManager (SAFETY tier).
    private readonly Dictionary<string, int> _priceFetchFailures = new();
    private const int MaxPriceFailures = 3;

    // R-3: Cache the market clock to avoid an Alpaca API call on every 30-second exit check cycle.
    // The clock is refreshed at most once per minute (60s TTL).
    private ClockInfo? _cachedClock;
    private DateTimeOffset _clockCacheTime = DateTimeOffset.MinValue;
    private static readonly TimeSpan ClockCacheTtl = TimeSpan.FromSeconds(60);

    // Scaling exits: track high water mark per symbol for trailing stop tiers
    private readonly Dictionary<string, decimal> _highWaterMarks = new();

    /// <summary>
    /// Main execution loop — run this in a hosted service background task.
    /// </summary>
    public virtual async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        logger.LogInformation("ExitManager starting with {interval}s check interval",
            _options.CheckIntervalSeconds);

        using var timer = new PeriodicTimer(
            TimeSpan.FromSeconds(_options.CheckIntervalSeconds));

        try
        {
            while (await timer.WaitForNextTickAsync(stoppingToken))
            {
                try
                {
                    var exitSignals = await CheckPositionsAsync(stoppingToken);
                    foreach (var signal in exitSignals)
                    {
                        // Publish FIRST, then set PendingExit so a bus failure doesn't phantom-lock position
                        _ = await eventBus.PublishAsync(signal, stoppingToken);

                        var pos = positionTracker.GetPosition(signal.Symbol);
                        if (pos != null)
                        {
                            pos.PendingExit = true;
                            await RecordExitAttemptAsync(signal.Symbol, stoppingToken);
                        }

                        logger.LogInformation(
                            "Published exit signal: {symbol} {reason} @ {price}",
                            signal.Symbol, signal.ExitReason, signal.ExitPrice);
                    }
                }
                catch (OperationCanceledException)
                {
                    throw;
                }
                catch (Exception ex)
                {
                    logger.LogError(ex, "Error in exit check cycle");
                }
            }
        }
        catch (OperationCanceledException)
        {
            logger.LogInformation("ExitManager stopped");
        }
    }

    /// <summary>
    /// Checks all positions for exit conditions.
    ///
    /// Skips positions with invalid ATR (cannot safely compute risk levels).
    /// ATR-based rules take full priority: when ATR is valid, fixed-% rules (3/5) are not evaluated.
    /// PendingExit is NOT set here — the caller (ExecuteAsync) sets it after successful publish.
    /// </summary>
    public async ValueTask<List<ExitSignalEvent>> CheckPositionsAsync(CancellationToken ct)
    {
        var signals = new List<ExitSignalEvent>();

        // R-3: Use cached clock if fresh (< 60s); otherwise fetch and cache a new one.
        // Reduces Alpaca clock API calls from once per 30s cycle to at most once per minute.
        ClockInfo clock;
        var now = DateTimeOffset.UtcNow;
        if (_cachedClock != null && now - _clockCacheTime < ClockCacheTtl)
        {
            clock = _cachedClock;
        }
        else
        {
            clock = await brokerService.GetClockAsync(ct);
            _cachedClock = clock;
            _clockCacheTime = now;
        }

        var positions = positionTracker.GetAllPositions();

        foreach (var (symbol, posData) in positions)
        {
            try
            {
                // Skip check if market closed (except 24/5 crypto)
                if (!clock.IsOpen && !_symbolClassifier.IsCrypto(symbol))
                {
                    continue;
                }

                // Skip if pending exit (waiting for order to fill/fail or in backoff period)
                if (posData.PendingExit)
                {
                    var nextRetryAt = await stateRepository.GetExitAttemptNextRetryAtAsync(symbol, ct);
                    // If no backoff recorded (null), order is still pending; keep waiting
                    // If backoff is recorded but not expired, keep waiting
                    if (nextRetryAt == null || now < nextRetryAt.Value)
                    {
                        continue; // Still waiting (no failure backoff, or backoff not yet expired)
                    }
                    // Backoff has expired; allow retry by clearing PendingExit
                    posData.PendingExit = false;
                }

                // Skip position if ATR is invalid — cannot safely compute risk levels
                if (posData.AtrValue <= 0 || !ValidateAtr(posData.AtrValue))
                {
                    logger.LogWarning(
                        "Invalid ATR for {symbol}: {atr}, skipping exit checks",
                        symbol, posData.AtrValue);
                    continue;
                }

                var currentPrice = await GetCurrentPriceAsync(symbol, ct);
                if (currentPrice <= 0)
                {
                    continue;
                }

                var atrDistanceMultiplier = 1.0m;
                if (volatilityRegimeDetector is { Enabled: true })
                {
                    var volRegime = await volatilityRegimeDetector.GetRegimeAsync(symbol, ct);
                    atrDistanceMultiplier = volRegime.StopMultiplier;
                    logger.LogDebug(
                        "Volatility stops for {symbol}: regime={regime} vol={vol:F6} bars={bars} stopMult={mult:F2}",
                        symbol, volRegime.Regime, volRegime.RealisedVolatility, volRegime.BarsInRegime, atrDistanceMultiplier);
                }

                // ATR levels are valid — compute once (atr_computed = true).
                // Fixed-% fallbacks (Rules 3 & 5) are mutually excluded when ATR is valid.
                var atrStop = posData.EntryPrice - (posData.AtrValue * _options.AtrStopLossMultiplier * atrDistanceMultiplier);
                var atrTarget = posData.EntryPrice + (posData.AtrValue * _options.AtrProfitTargetMultiplier * atrDistanceMultiplier);

                // Rule 1: ATR-based stop loss (highest priority)
                if (currentPrice <= atrStop)
                {
                    signals.Add(new ExitSignalEvent(symbol, "ATR_STOP_LOSS", currentPrice, DateTimeOffset.UtcNow));
                    continue;
                }

                // Rule 2: ATR-based profit target
                if (currentPrice >= atrTarget)
                {
                    signals.Add(new ExitSignalEvent(symbol, "ATR_PROFIT_TARGET", currentPrice, DateTimeOffset.UtcNow));
                    continue;
                }

                // Rule 4: Trailing stop (independent of ATR — always active when set)
                if (posData.TrailingStopPrice > 0 && currentPrice <= posData.TrailingStopPrice)
                {
                    signals.Add(new ExitSignalEvent(symbol, "TRAILING_STOP", currentPrice, DateTimeOffset.UtcNow));
                    continue;
                }

                // Scaling exits: check for partial profit-taking at multiple tiers
                if (_options.ScalingExits.Count > 0)
                {
                    var scalingExits = await CheckScalingExitsAsync(symbol, posData, currentPrice, ct);
                    foreach (var (signal, _) in scalingExits)
                    {
                        signals.Add(signal);
                    }
                }

                // Rules 3 & 5 (fixed-% stop/profit) intentionally omitted when ATR is valid.
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "Error checking exit conditions for {symbol}", symbol);
            }
        }

        return signals;
    }

    /// <summary>
    /// Handles order update events: clears pending_exit flag on terminal failure.
    /// </summary>
    public async ValueTask HandleOrderUpdateAsync(OrderUpdateEvent orderUpdate, CancellationToken ct)
    {
        var terminalFailureStates = new[]
        {
            OrderState.Canceled,
            OrderState.Expired,
            OrderState.Rejected,
        };

        if (!terminalFailureStates.Contains(orderUpdate.Status))
        {
            return;
        }

        var positions = positionTracker.GetAllPositions();
        foreach (var (symbol, posData) in positions)
        {
            if (posData.PendingExit && symbol == orderUpdate.Symbol)
            {
                posData.PendingExit = false;
                await RecordExitAttemptFailureAsync(symbol, ct);
                logger.LogWarning(
                    "Cleared pending exit for {symbol} after order {status}",
                    symbol, orderUpdate.Status);
                break;
            }
        }
    }

    /// <summary>
    /// Validates ATR is a finite positive number.
    /// </summary>
    private static bool ValidateAtr(decimal atr) =>
        atr > 0 && atr != decimal.MinValue && atr != decimal.MaxValue;

    /// <summary>
    /// Gets current mid-price for symbol from market data snapshot.
    /// Tracks consecutive failures per symbol; after MaxPriceFailures consecutive failures
    /// the market_data_degraded KV flag is set (blocks new entries in RiskManager SAFETY tier).
    /// On success, clears that symbol's failure counter and clears the KV flag when all symbols OK.
    /// Returns 0m on individual failure so the position is skipped this cycle.
    /// </summary>
    private async ValueTask<decimal> GetCurrentPriceAsync(string symbol, CancellationToken ct)
    {
        try
        {
            var snapshot = await marketDataClient.GetSnapshotAsync(symbol, ct);

            // Success: reset this symbol's failure counter.
            if (_priceFetchFailures.TryGetValue(symbol, out var prev) && prev > 0)
            {
                _priceFetchFailures[symbol] = 0;

                // Clear the degraded flag only when every *currently held* position has a
                // clean price feed. Counters for symbols that are no longer held are stale
                // and must not prevent recovery — prune them here.
                var currentSymbols = positionTracker.GetAllPositions().Keys.ToHashSet();
                foreach (var key in _priceFetchFailures.Keys.Where(k => !currentSymbols.Contains(k)).ToList())
                    _priceFetchFailures.Remove(key);

                if (_priceFetchFailures.Values.All(v => v == 0))
                {
                    await stateRepository.SetStateAsync("market_data_degraded", "false", ct);
                    logger.LogInformation("Market data recovered for all held positions; cleared market_data_degraded flag");
                }
            }

            return snapshot.MidPrice;
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to get current price for {symbol}", symbol);
            _priceFetchFailures.TryGetValue(symbol, out var failures);
            _priceFetchFailures[symbol] = failures + 1;

            if (_priceFetchFailures[symbol] >= MaxPriceFailures)
            {
                logger.LogError(
                    "Price fetch failed {count} consecutive times for {symbol}; setting market_data_degraded=true to block new entries",
                    _priceFetchFailures[symbol], symbol);
                await stateRepository.SetStateAsync("market_data_degraded", "true", ct);
            }

            return 0m;
        }
    }

    private async ValueTask RecordExitAttemptAsync(string symbol, CancellationToken ct)
    {
        try
        {
            await stateRepository.RecordExitAttemptAsync(symbol, ct);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to record exit attempt for {symbol}", symbol);
        }
    }

    private async ValueTask RecordExitAttemptFailureAsync(string symbol, CancellationToken ct)
    {
        try
        {
            await stateRepository.RecordExitAttemptFailureAsync(symbol, ct);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to record exit failure for {symbol}", symbol);
        }
    }

    /// <summary>
    /// Checks for scaling exit conditions (partial profit-taking at multiple tiers).
    /// Only runs if ScalingExits is configured and not empty.
    /// Returns list of scaling exit signals (each with quantity to close).
    /// </summary>
    private async ValueTask<List<(ExitSignalEvent Signal, decimal Quantity)>> CheckScalingExitsAsync(
        string symbol,
        PositionData posData,
        decimal currentPrice,
        CancellationToken ct)
    {
        var results = new List<(ExitSignalEvent, decimal)>();

        if (_options.ScalingExits.Count == 0)
            return results;

        try
        {
            // Update high water mark for this symbol
            if (!_highWaterMarks.TryGetValue(symbol, out var hwm) || currentPrice > hwm)
            {
                _highWaterMarks[symbol] = currentPrice;
            }

            var atrDistanceMultiplier = 1.0m;
            if (volatilityRegimeDetector is { Enabled: true })
            {
                var volRegime = await volatilityRegimeDetector.GetRegimeAsync(symbol, ct);
                atrDistanceMultiplier = volRegime.StopMultiplier;
            }

            foreach (var tier in _options.ScalingExits.OrderBy(t => t.DistanceMultiplier))
            {
                if (tier.Trigger.Equals("ProfitTarget", StringComparison.OrdinalIgnoreCase))
                {
                    var targetPrice = posData.EntryPrice + (posData.AtrValue * tier.DistanceMultiplier * atrDistanceMultiplier);
                    if (currentPrice >= targetPrice)
                    {
                        var qtyToClose = _symbolClassifier.IsCrypto(symbol)
                            ? Math.Max(0.0001m, Math.Round(posData.CurrentQuantity * tier.PercentageToClose, 8))
                            : Math.Max(1m, Math.Floor(posData.CurrentQuantity * tier.PercentageToClose));

                        if (qtyToClose > 0)
                        {
                            var signal = new ExitSignalEvent(
                                symbol,
                                $"SCALING_PROFIT_TIER_{tier.DistanceMultiplier:F1}x",
                                currentPrice,
                                DateTimeOffset.UtcNow);
                            results.Add((signal, qtyToClose));

                            logger.LogInformation(
                                "Scaling exit triggered: {symbol} SELL {qty} shares @ {price:F2} (tier {mult}x ATR, profit target)",
                                symbol, qtyToClose, currentPrice, tier.DistanceMultiplier);
                        }
                    }
                }
                else if (tier.Trigger.Equals("TrailingStop", StringComparison.OrdinalIgnoreCase))
                {
                    var hwPrice = _highWaterMarks.GetValueOrDefault(symbol, posData.EntryPrice);
                    var trailDistance = posData.AtrValue * tier.DistanceMultiplier * atrDistanceMultiplier;
                    var trailPrice = hwPrice - trailDistance;

                    if (currentPrice <= trailPrice)
                    {
                        var qtyToClose = _symbolClassifier.IsCrypto(symbol)
                            ? Math.Max(0.0001m, Math.Round(posData.CurrentQuantity * tier.PercentageToClose, 8))
                            : Math.Max(1m, Math.Floor(posData.CurrentQuantity * tier.PercentageToClose));

                        if (qtyToClose > 0)
                        {
                            var signal = new ExitSignalEvent(
                                symbol,
                                $"SCALING_TRAIL_TIER_{tier.DistanceMultiplier:F1}x",
                                currentPrice,
                                DateTimeOffset.UtcNow);
                            results.Add((signal, qtyToClose));

                            logger.LogInformation(
                                "Scaling exit triggered: {symbol} SELL {qty} shares @ {price:F2} (tier {mult}x ATR, trailing stop)",
                                symbol, qtyToClose, currentPrice, tier.DistanceMultiplier);
                        }
                    }
                }
            }
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Error checking scaling exits for {symbol}", symbol);
        }

        return results;
    }
}
